libs/README.md

```diff
@@ -0,0 +1,375 @@
+break simply because we move the buttons to other places.
+
+Another potential problem is the destruction of the main window since
+the frontend part controls the backend. If the main window is destroyed
+accidentally by the system, we lose the state, and the containers
+become unmanaged.
+
+It is also impossible to create a multi-window interface, or
+a container that manages others (a kind of launchpad for servers).
+
+Another class of potential problems is related to the fact that we,
+for example, send SQL queries directly from the frontend to the backend.
+And, in theory, we have to cover that requests with tests as well.
+
+### The frontend
+
+The frontend uses `React` framework and `Redux` library.
+
+The `Redux` maintains the main state of the app and calls API methods
+that we exported from the backend.
+
+## The improved architecture
+
+The approach discussed below requires a lot of effort, but it can
+significantly increase the reliability of the launchpad application.
+
+The idea is to move all control functions to the backend and transfer
+the application state and changes to this state to the frontend using
+Tauri events.
+
+In this case, in general, you will not need to export your API methods
+since the application will also be controlled by sending messages from
+frontend to backend, which will launch processes that change that
+mentioned above - shared state.
+
+### Events channels
+
+To connect UI to the new engine we should use two calls on the frontend that is
+the part of the Tauri API.
+
+To subscribe to the state and changes of that we should subscribe to the
+`tari://reactions` stream of events:
+
+```js
+const unlisten = await listen('tari://reactions', (event) => { })
+```
+
+And to send a control action we should send an event to the `tari://actions`
+stream:
+
+```js
+function sendEvent(action) {
+    emit('tari://actions', action)
+}
+```
+
+The reasonable questions may occur: if we send actions and have to wait
+for the state to be updated by a backend and a change event will be sent
+and applied to the state in the UI, will we have a significant delay
+noticeable to the user?
+
+The answer is no; we don't have too long delays because it's
+a desktop app, and it takes up to 2ms to send an event from
+the backend to a server.
+
+### The state
+
+The state is the main object exists in the backend and has to be
+copied to the frontend.
+
+The state is a struct with:
+
+- a configuration of the app
+- a map of the states of the containers
+- a state of the wallet
+
+```rust
+pub struct LaunchpadState {
+    pub config: Option<LaunchpadConfig>,
+    pub containers: HashMap<TaskId, TaskState>,
+    pub wallet: WalletState,
+}
+```
+
+The state above should be declared in the frontend part and received from the
+`tari://reactions` stream that send events of the following `Reaction` type:
+
+```rust
+pub enum Reaction {
+    State(LaunchpadState),
+    Delta(LaunchpadDelta),
+}
```

Comment:
Ok, so the first message (or when re-requested) would be a `State` message to set the state. Subsequent messages will mostly be `Delta` messages which the FE FSM can use to decide what UI changes to make and/or which actions to send.

There are some further elaborations below which elucidate a bit more.